// @claude-flow/claims/dist/api/cli-types.js
/**
 * CLI Type Stubs for Claims Module
 *
 * Local type definitions to avoid cross-package imports.
 * These mirror the types from @claude-flow/cli for use in claims commands.
 */
// =============================================================================
// Output Utilities (mirrors @claude-flow/cli/src/output.ts)
// =============================================================================
export const output = {
    log: (message) => {
        console.log(message);
    },
    error: (message) => {
        console.error(`Error: ${message}`);
    },
    warn: (message) => {
        console.warn(`Warning: ${message}`);
    },
    warning: (message) => {
        return `⚠ ${message}`;
    },
    success: (message) => {
        console.log(`✓ ${message}`);
    },
    info: (message) => {
        console.log(`ℹ ${message}`);
    },
    table: (data) => {
        console.table(data);
    },
    json: (data) => {
        console.log(JSON.stringify(data, null, 2));
    },
    // Formatting helpers that return strings for composition
    dim: (message) => message,
    bold: (message) => message,
    italic: (message) => message,
    highlight: (message) => message,
    code: (message) => `\`${message}\``,
    link: (url, text) => text ? `${text} (${url})` : url,
    list: (items) => items.map(i => `  • ${i}`).join('\n'),
    header: (message) => `\n${message}\n${'─'.repeat(message.length)}`,
    // Colors
    red: (message) => message,
    green: (message) => message,
    yellow: (message) => message,
    blue: (message) => message,
    cyan: (message) => message,
    magenta: (message) => message,
    gray: (message) => message,
    white: (message) => message,
};
export async function select(message, options) {
    // In a real implementation, this would use a terminal prompt library
    // For now, return the first option as a stub
    console.log(`[Prompt] ${message}`);
    options.forEach((opt, i) => console.log(`  ${i + 1}. ${opt.label}`));
    return options[0]?.value;
}
export async function confirm(message, defaultValue = false) {
    console.log(`[Confirm] ${message} (default: ${defaultValue ? 'yes' : 'no'})`);
    return defaultValue;
}
export async function input(message, defaultValue = '') {
    console.log(`[Input] ${message} (default: ${defaultValue})`);
    return defaultValue;
}
// =============================================================================
// MCP Client Utilities (mirrors @claude-flow/cli/src/mcp-client.ts)
// =============================================================================
export class MCPClientError extends Error {
    code;
    details;
    constructor(message, code, details) {
        super(message);
        this.code = code;
        this.details = details;
        this.name = 'MCPClientError';
    }
}
export async function callMCPTool(toolName, params) {
    // MCP tool call - delegates to active MCP server
    console.log(`[MCP] Calling tool: ${toolName}`, params);
    return { success: true, data: {} };
}
