// @claude-flow/security/dist/plugins/integrity-verifier.js
/**
 * PluginIntegrityVerifier — install-layer security for plugin supply chain.
 *
 * Implements P1 of ADR-145 (ruvnet/ruflo#2254): Stage-1 Ed25519 signature
 * verification at `plugins install`. Stage-2 (semantic-intent scan against
 * SCH attacks) lands in P2; per-namespace write ACLs land in P3-P4.
 *
 * Threat model
 * ------------
 * Ruflo's plugin install path fetches manifests from IPFS via Pinata with
 * NO signature verification and NO intent analysis. Two attacks land today:
 *
 *   - DDIPE (arXiv:2604.03081): malicious logic embedded in plugin docs and
 *     config templates. 11.6-33.5% bypass rate across 4 frameworks/5 models;
 *     2.5% evade both detection and alignment. Stage-1 signing blocks the
 *     static-payload variants.
 *
 *   - SCH (arXiv:2605.14460): malicious intent wrapped as natural-language
 *     "compliance rules" in plugin descriptions. The agent generates the
 *     harmful code at runtime — no static payload exists. 77.67% breach
 *     success, 0.00% scanner detection. Stage-2 (P2) catches this; Stage-1
 *     does not.
 *
 * Scope (P1)
 * ----------
 * - Ed25519 signature verification of a detached signature over the manifest.
 * - Trust-anchor allowlist: the publisher's signing key fingerprint must be
 *   in `trust-anchors.json`, gated on CODEOWNERS review.
 * - Backwards compatible: default mode is warn-only. `CLAUDE_FLOW_STRICT_PLUGINS=true`
 *   makes verification a hard gate.
 *
 * Non-goals (P1)
 * --------------
 * - Semantic intent scan (SCH defence). Lands in P2 with a pattern-rule
 *   fallback for environments without LLM credentials.
 * - Sandboxing the plugin runtime. Orthogonal blast-radius concern.
 *
 * Reference: ADR-145, arXiv:2605.14460 (SCH), arXiv:2604.03081 (DDIPE),
 * arXiv:2604.16548 (Mnemonic Sovereignty survey).
 */
import { createHash } from 'node:crypto';
import * as nodeCrypto from 'node:crypto';
/**
 * Canonical JSON serialisation for hashing. Deterministic key ordering at
 * every nesting level. Pure; identical input → identical output bytes.
 */
export function canonicalize(value) {
    if (value === null || typeof value !== 'object')
        return JSON.stringify(value);
    if (Array.isArray(value))
        return '[' + value.map(canonicalize).join(',') + ']';
    const obj = value;
    const keys = Object.keys(obj).sort();
    return '{' + keys.map(k => JSON.stringify(k) + ':' + canonicalize(obj[k])).join(',') + '}';
}
/** Hex-encoded SHA-256 of the canonical-JSON manifest. */
export function hashManifest(manifest) {
    return createHash('sha256').update(canonicalize(manifest)).digest('hex');
}
/** Short fingerprint for telemetry — first 16 hex chars of the key. */
export function fingerprint(publicKey) {
    return publicKey.slice(0, 16);
}
/**
 * Find the trust anchor that vouches for a manifest. Returns the matched
 * anchor (with anchor.publicKey === signer.publicKey) or null.
 *
 * Scope-matching uses a minimal glob: `*` at the end of the scope string
 * matches any suffix. Empty/missing scope matches every plugin id.
 */
export function findAnchor(anchors, pluginId, signerPublicKey, now) {
    for (const a of anchors) {
        if (a.publicKey !== signerPublicKey)
            continue;
        if (a.expiresAt && new Date(a.expiresAt).getTime() <= now)
            continue;
        if (!a.scope || a.scope === '*' || a.scope === pluginId)
            return a;
        if (a.scope.endsWith('*') && pluginId.startsWith(a.scope.slice(0, -1)))
            return a;
    }
    return null;
}
/**
 * `PluginIntegrityVerifier` — Stage-1 verifier (signature + trust anchor).
 *
 * Pure-ish: holds the trust-anchor list and clock, no other state. Safe to
 * construct per-invocation or share.
 */
export class PluginIntegrityVerifier {
    config;
    constructor(config) {
        this.config = config;
    }
    /**
     * Verify a signed manifest. Returns the verification status; the caller
     * decides what to do with `signature-missing` / `manifest-hash-mismatch`
     * etc. based on its strict-mode policy.
     */
    async verify(signed) {
        const now = (this.config.now ?? Date.now)();
        const pluginId = signed.manifest?.id ?? '<unknown>';
        if (!signed.signature || !signed.publicKey) {
            return { status: 'signature-missing', pluginId, ts: now };
        }
        // Re-hash the manifest and compare to the signed hash.
        const recomputed = hashManifest(signed.manifest);
        if (recomputed !== signed.manifestHash) {
            return {
                status: 'manifest-hash-mismatch',
                pluginId,
                signerFingerprint: fingerprint(signed.publicKey),
                ts: now,
                detail: `expected ${signed.manifestHash}, got ${recomputed}`,
            };
        }
        // Find a trust anchor that vouches for this signer + plugin.
        const anchor = findAnchor(this.config.trustAnchors.anchors, pluginId, signed.publicKey, now);
        if (!anchor) {
            return {
                status: 'unknown-signer',
                pluginId,
                signerFingerprint: fingerprint(signed.publicKey),
                ts: now,
            };
        }
        // Ed25519 verify. Probe @noble/ed25519 (already a workspace dep); fall
        // back to the failure status if it isn't resolvable in this environment
        // (mirrors verify.mjs's #1880 precondition pattern).
        const ed = await loadEd25519();
        if (!ed) {
            return {
                status: 'signature-invalid',
                pluginId,
                signerFingerprint: fingerprint(signed.publicKey),
                ts: now,
                detail: '@noble/ed25519 not available — cannot verify',
            };
        }
        // Required by @noble/ed25519 v2 for sync sha512.
        if (!ed.etc.sha512Sync) {
            ed.etc.sha512Sync = (...m) => {
                const h = nodeCrypto.createHash('sha512');
                for (const x of m)
                    h.update(x);
                return h.digest();
            };
        }
        let ok = false;
        try {
            const sigBytes = hexToBytes(signed.signature);
            const msgBytes = hexToBytes(signed.manifestHash);
            const pubBytes = hexToBytes(signed.publicKey);
            ok = await ed.verify(sigBytes, msgBytes, pubBytes);
        }
        catch {
            ok = false;
        }
        if (!ok) {
            return {
                status: 'signature-invalid',
                pluginId,
                signerFingerprint: fingerprint(signed.publicKey),
                ts: now,
            };
        }
        return { status: 'pass', pluginId, signerFingerprint: fingerprint(signed.publicKey), ts: now };
    }
}
// ─── helpers ────────────────────────────────────────────────────────────
function hexToBytes(hex) {
    const out = new Uint8Array(hex.length >>> 1);
    for (let i = 0; i < out.length; i++)
        out[i] = parseInt(hex.substr(i * 2, 2), 16);
    return out;
}
async function loadEd25519() {
    try {
        const mod = await import('@noble/ed25519');
        return mod;
    }
    catch {
        return null;
    }
}
