// @claude-flow/claims/dist/infrastructure/claim-repository.js
/**
 * @claude-flow/claims - Claim Repository Implementation
 * SQLite-based persistence for claims (ADR-016)
 *
 * @module v3/claims/infrastructure/claim-repository
 */
// =============================================================================
// In-Memory Claim Repository (Default Implementation)
// =============================================================================
/**
 * In-memory implementation of the claim repository
 * Suitable for development and testing
 */
export class InMemoryClaimRepository {
    claims = new Map();
    issueIndex = new Map(); // issueId:repo -> claimId
    claimantIndex = new Map(); // claimantId -> claimIds
    async initialize() {
        // No initialization needed for in-memory store
    }
    async shutdown() {
        this.claims.clear();
        this.issueIndex.clear();
        this.claimantIndex.clear();
    }
    // ==========================================================================
    // CRUD Operations
    // ==========================================================================
    async save(claim) {
        const fullClaim = this.ensureFullClaim(claim);
        this.claims.set(claim.id, fullClaim);
        // Update indexes
        const issueKey = this.getIssueKey(claim.issueId, claim.repository ?? '');
        this.issueIndex.set(issueKey, claim.id);
        const claimantId = claim.claimant.id;
        if (!this.claimantIndex.has(claimantId)) {
            this.claimantIndex.set(claimantId, new Set());
        }
        this.claimantIndex.get(claimantId).add(claim.id);
    }
    async update(claim) {
        await this.save(claim);
    }
    async findById(claimId) {
        return this.claims.get(claimId) ?? null;
    }
    async findByIssueId(issueId, repository) {
        const issueKey = this.getIssueKey(issueId, repository ?? '');
        const claimId = this.issueIndex.get(issueKey);
        if (!claimId)
            return null;
        const claim = this.claims.get(claimId);
        if (!claim)
            return null;
        // Return only if claim is active
        if (this.isActiveStatus(claim.status)) {
            return claim;
        }
        return null;
    }
    async findByClaimant(claimant) {
        return this.findByAgentId(claimant.id);
    }
    async findByAgentId(agentId) {
        const claimIds = this.claimantIndex.get(agentId);
        if (!claimIds)
            return [];
        return Array.from(claimIds)
            .map((id) => this.claims.get(id))
            .filter((c) => c !== undefined);
    }
    async findByStatus(status) {
        return Array.from(this.claims.values()).filter((c) => c.status === status);
    }
    async findStealable(agentType) {
        return Array.from(this.claims.values()).filter((c) => {
            if (c.status !== 'stealable')
                return false;
            if (!c.stealInfo)
                return true;
            if (!agentType)
                return true;
            if (!c.stealInfo.allowedStealerTypes)
                return true;
            return c.stealInfo.allowedStealerTypes.includes(agentType);
        });
    }
    async findContested() {
        return Array.from(this.claims.values()).filter((c) => c.contestInfo && !c.contestInfo.resolution);
    }
    async findAll() {
        return Array.from(this.claims.values());
    }
    async delete(claimId) {
        const claim = this.claims.get(claimId);
        if (claim) {
            // Remove from indexes
            const issueKey = this.getIssueKey(claim.issueId, claim.repository ?? '');
            this.issueIndex.delete(issueKey);
            const claimantId = claim.claimant.id;
            const claimantClaims = this.claimantIndex.get(claimantId);
            if (claimantClaims) {
                claimantClaims.delete(claimId);
                if (claimantClaims.size === 0) {
                    this.claimantIndex.delete(claimantId);
                }
            }
            this.claims.delete(claimId);
        }
    }
    // ==========================================================================
    // Query Operations
    // ==========================================================================
    async findActiveClaims() {
        return Array.from(this.claims.values()).filter((c) => this.isActiveStatus(c.status));
    }
    async findStaleClaims(staleSince) {
        const staleTimestamp = staleSince.getTime();
        return Array.from(this.claims.values()).filter((c) => this.isActiveStatus(c.status) &&
            c.lastActivityAt.getTime() < staleTimestamp);
    }
    async findClaimsWithPendingHandoffs() {
        return Array.from(this.claims.values()).filter((c) => c.status === 'pending_handoff');
    }
    async countByClaimant(claimantId) {
        return this.claimantIndex.get(claimantId)?.size ?? 0;
    }
    async countByAgentId(agentId) {
        const claims = await this.findByAgentId(agentId);
        return claims.filter((c) => this.isActiveStatus(c.status)).length;
    }
    // ==========================================================================
    // Extended Query Operations (ADR-016)
    // ==========================================================================
    async query(options) {
        let results = Array.from(this.claims.values());
        // Apply filters
        if (options.claimantId) {
            results = results.filter((c) => c.claimant.id === options.claimantId);
        }
        if (options.claimantType) {
            results = results.filter((c) => c.claimant.type === options.claimantType);
        }
        if (options.status) {
            const statuses = Array.isArray(options.status) ? options.status : [options.status];
            results = results.filter((c) => statuses.includes(c.status));
        }
        if (options.repository) {
            results = results.filter((c) => c.repository === options.repository);
        }
        if (options.issueId) {
            results = results.filter((c) => c.issueId === options.issueId);
        }
        if (options.stealableOnly) {
            results = results.filter((c) => c.status === 'stealable');
        }
        if (options.blockedOnly) {
            results = results.filter((c) => c.blockedReason !== undefined);
        }
        if (options.createdAfter) {
            results = results.filter((c) => c.claimedAt.getTime() >= options.createdAfter);
        }
        if (options.createdBefore) {
            results = results.filter((c) => c.claimedAt.getTime() <= options.createdBefore);
        }
        if (options.updatedAfter) {
            results = results.filter((c) => c.lastActivityAt.getTime() >= options.updatedAfter);
        }
        // Apply sorting
        if (options.sortBy) {
            results.sort((a, b) => {
                let aVal, bVal;
                switch (options.sortBy) {
                    case 'claimedAt':
                        aVal = a.claimedAt.getTime();
                        bVal = b.claimedAt.getTime();
                        break;
                    case 'updatedAt':
                        aVal = a.lastActivityAt.getTime();
                        bVal = b.lastActivityAt.getTime();
                        break;
                    case 'progress':
                        aVal = a.progress;
                        bVal = b.progress;
                        break;
                    default:
                        return 0;
                }
                return options.sortDirection === 'desc' ? bVal - aVal : aVal - bVal;
            });
        }
        // Apply pagination
        if (options.offset) {
            results = results.slice(options.offset);
        }
        if (options.limit) {
            results = results.slice(0, options.limit);
        }
        return results;
    }
    async getStatistics() {
        const claims = Array.from(this.claims.values());
        const byStatus = {
            active: 0,
            paused: 0,
            'handoff-pending': 0,
            'review-requested': 0,
            blocked: 0,
            stealable: 0,
            completed: 0,
        };
        const byClaimantType = {
            human: 0,
            agent: 0,
        };
        const byRepository = {};
        let totalDuration = 0;
        let completedCount = 0;
        let totalProgress = 0;
        const now = Date.now();
        const last24h = now - 24 * 60 * 60 * 1000;
        for (const claim of claims) {
            byStatus[claim.status] = (byStatus[claim.status] ?? 0) + 1;
            byClaimantType[claim.claimant.type] = (byClaimantType[claim.claimant.type] ?? 0) + 1;
            const repo = claim.repository ?? 'unknown';
            byRepository[repo] = (byRepository[repo] ?? 0) + 1;
            totalProgress += claim.progress;
            if (claim.status === 'completed') {
                completedCount++;
                const duration = claim.lastActivityAt.getTime() - claim.claimedAt.getTime();
                totalDuration += duration;
            }
        }
        const completedLast24h = claims.filter((c) => c.status === 'completed' && c.lastActivityAt.getTime() >= last24h).length;
        return {
            totalClaims: claims.length,
            byStatus: byStatus,
            byPriority: { critical: 0, high: 0, medium: 0, low: 0 },
            byClaimantType: byClaimantType,
            avgDurationMs: completedCount > 0 ? totalDuration / completedCount : 0,
            avgProgress: claims.length > 0 ? totalProgress / claims.length : 0,
            activeSteals: claims.filter((c) => c.status === 'stealable').length,
            pendingHandoffs: claims.filter((c) => c.status === 'pending_handoff').length,
            completedLast24h,
            byRepository,
        };
    }
    // ==========================================================================
    // Helper Methods
    // ==========================================================================
    getIssueKey(issueId, repository) {
        return `${repository}:${issueId}`;
    }
    isActiveStatus(status) {
        return ['active', 'paused', 'blocked', 'pending_handoff', 'in_review', 'stealable'].includes(status);
    }
    ensureFullClaim(claim) {
        const fullClaim = claim;
        if (fullClaim.progress === undefined) {
            fullClaim.progress = 0;
        }
        return fullClaim;
    }
}
// =============================================================================
// Factory Function
// =============================================================================
/**
 * Create a new claim repository
 */
export function createClaimRepository() {
    return new InMemoryClaimRepository();
}
