// @claude-flow/claims/dist/infrastructure/event-store.js
/**
 * @claude-flow/claims - Event Store Implementation
 * Event sourcing storage for claims (ADR-007, ADR-016)
 *
 * @module v3/claims/infrastructure/event-store
 */
// =============================================================================
// In-Memory Event Store Implementation
// =============================================================================
/**
 * In-memory implementation of the event store
 * Suitable for development and testing
 */
export class InMemoryClaimEventStore {
    events = [];
    aggregateVersions = new Map();
    subscriptions = [];
    nextSubscriptionId = 0;
    async initialize() {
        // No initialization needed for in-memory store
    }
    async shutdown() {
        this.events = [];
        this.aggregateVersions.clear();
        this.subscriptions = [];
    }
    // ==========================================================================
    // Write Operations
    // ==========================================================================
    async append(event) {
        // Assign version
        const currentVersion = this.aggregateVersions.get(event.aggregateId) ?? 0;
        const newVersion = currentVersion + 1;
        event.version = newVersion;
        // Store event
        this.events.push(event);
        this.aggregateVersions.set(event.aggregateId, newVersion);
        // Notify subscribers
        await this.notifySubscribers(event);
    }
    async appendBatch(events) {
        for (const event of events) {
            await this.append(event);
        }
    }
    // ==========================================================================
    // Read Operations
    // ==========================================================================
    async getEvents(claimId, fromVersion) {
        return this.events.filter((e) => e.aggregateId === claimId &&
            (fromVersion === undefined || e.version >= fromVersion));
    }
    async getEventsByType(type) {
        return this.events.filter((e) => e.type === type);
    }
    async getEventsByIssueId(issueId) {
        return this.events.filter((e) => e.payload?.issueId === issueId);
    }
    async query(filter) {
        let results = [...this.events];
        if (filter.aggregateId) {
            results = results.filter((e) => e.aggregateId === filter.aggregateId);
        }
        if (filter.eventTypes && filter.eventTypes.length > 0) {
            results = results.filter((e) => filter.eventTypes.includes(e.type));
        }
        if (filter.fromTimestamp !== undefined) {
            results = results.filter((e) => e.timestamp >= filter.fromTimestamp);
        }
        if (filter.toTimestamp !== undefined) {
            results = results.filter((e) => e.timestamp <= filter.toTimestamp);
        }
        if (filter.fromVersion !== undefined) {
            results = results.filter((e) => e.version >= filter.fromVersion);
        }
        if (filter.toVersion !== undefined) {
            results = results.filter((e) => e.version <= filter.toVersion);
        }
        // Apply pagination
        if (filter.offset) {
            results = results.slice(filter.offset);
        }
        if (filter.limit) {
            results = results.slice(0, filter.limit);
        }
        return results;
    }
    // ==========================================================================
    // Subscription Operations
    // ==========================================================================
    subscribe(eventTypes, handler) {
        const subscription = {
            id: `sub-${++this.nextSubscriptionId}`,
            eventTypes,
            handler,
        };
        this.subscriptions.push(subscription);
        // Return unsubscribe function
        return () => {
            const index = this.subscriptions.findIndex((s) => s.id === subscription.id);
            if (index !== -1) {
                this.subscriptions.splice(index, 1);
            }
        };
    }
    subscribeAll(handler) {
        return this.subscribe([], handler);
    }
    async notifySubscribers(event) {
        for (const subscription of this.subscriptions) {
            // If no event types specified, handler receives all events
            if (subscription.eventTypes.length === 0 ||
                subscription.eventTypes.includes(event.type)) {
                try {
                    await subscription.handler(event);
                }
                catch (error) {
                    console.error(`Event handler error for subscription ${subscription.id}:`, error);
                }
            }
        }
    }
    // ==========================================================================
    // Aggregate Operations
    // ==========================================================================
    async getAggregateVersion(aggregateId) {
        return this.aggregateVersions.get(aggregateId) ?? 0;
    }
    async getAggregateState(aggregateId, reducer, initialState) {
        const events = await this.getEvents(aggregateId);
        return events.reduce((state, event) => reducer(state, event), initialState);
    }
    // ==========================================================================
    // Snapshot Operations
    // ==========================================================================
    snapshots = new Map();
    async saveSnapshot(aggregateId, state, version) {
        this.snapshots.set(aggregateId, { state, version });
    }
    async getSnapshot(aggregateId) {
        const snapshot = this.snapshots.get(aggregateId);
        if (!snapshot)
            return null;
        return snapshot;
    }
    async getStateFromSnapshot(aggregateId, reducer, initialState) {
        // Try to get snapshot
        const snapshot = await this.getSnapshot(aggregateId);
        let state;
        let fromVersion;
        if (snapshot) {
            state = snapshot.state;
            fromVersion = snapshot.version + 1;
        }
        else {
            state = initialState;
            fromVersion = 1;
        }
        // Apply events after snapshot
        const events = await this.getEvents(aggregateId, fromVersion);
        return events.reduce((s, event) => reducer(s, event), state);
    }
    // ==========================================================================
    // Statistics
    // ==========================================================================
    async getEventCount() {
        return this.events.length;
    }
    async getEventCountByType() {
        const counts = {};
        for (const event of this.events) {
            counts[event.type] = (counts[event.type] ?? 0) + 1;
        }
        return counts;
    }
    async getAggregateCount() {
        return this.aggregateVersions.size;
    }
}
// =============================================================================
// Factory Function
// =============================================================================
/**
 * Create a new event store
 */
export function createClaimEventStore() {
    return new InMemoryClaimEventStore();
}
