// @ruvector/ruvllm/dist/cjs/benchmarks/index.js
"use strict";
/**
 * RuvLTRA Benchmark Suite
 *
 * Comprehensive benchmarks for evaluating RuvLTRA models
 * on Claude Code-specific use cases.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLUSTER_TEST_CASES = exports.SEARCH_TEST_CASES = exports.SIMILARITY_TEST_PAIRS = exports.ROUTING_TEST_CASES = void 0;
exports.runFullBenchmark = runFullBenchmark;
exports.formatFullResults = formatFullResults;
exports.compareModels = compareModels;
__exportStar(require("./routing-benchmark"), exports);
__exportStar(require("./embedding-benchmark"), exports);
__exportStar(require("./model-comparison"), exports);
const routing_benchmark_1 = require("./routing-benchmark");
Object.defineProperty(exports, "ROUTING_TEST_CASES", { enumerable: true, get: function () { return routing_benchmark_1.ROUTING_TEST_CASES; } });
const embedding_benchmark_1 = require("./embedding-benchmark");
Object.defineProperty(exports, "SIMILARITY_TEST_PAIRS", { enumerable: true, get: function () { return embedding_benchmark_1.SIMILARITY_TEST_PAIRS; } });
Object.defineProperty(exports, "SEARCH_TEST_CASES", { enumerable: true, get: function () { return embedding_benchmark_1.SEARCH_TEST_CASES; } });
Object.defineProperty(exports, "CLUSTER_TEST_CASES", { enumerable: true, get: function () { return embedding_benchmark_1.CLUSTER_TEST_CASES; } });
/**
 * Run all benchmarks with a given model
 */
function runFullBenchmark(router, embedder, similarityFn, modelName = 'unknown') {
    const routing = (0, routing_benchmark_1.runRoutingBenchmark)(router);
    const embedding = (0, embedding_benchmark_1.runEmbeddingBenchmark)(embedder, similarityFn);
    return {
        routing,
        embedding,
        timestamp: new Date().toISOString(),
        model: modelName,
    };
}
/**
 * Format full benchmark results
 */
function formatFullResults(results) {
    const lines = [];
    lines.push('');
    lines.push('╔═══════════════════════════════════════════════════════════════════════════╗');
    lines.push('║                    RUVLTRA BENCHMARK SUITE                                ║');
    lines.push('║            Claude Code Use Case Evaluation                                ║');
    lines.push('╠═══════════════════════════════════════════════════════════════════════════╣');
    lines.push(`║  Model: ${results.model.padEnd(64)}║`);
    lines.push(`║  Date:  ${results.timestamp.padEnd(64)}║`);
    lines.push('╚═══════════════════════════════════════════════════════════════════════════╝');
    lines.push((0, routing_benchmark_1.formatRoutingResults)(results.routing));
    lines.push((0, embedding_benchmark_1.formatEmbeddingResults)(results.embedding));
    // Overall assessment
    lines.push('');
    lines.push('═══════════════════════════════════════════════════════════════');
    lines.push('                      OVERALL ASSESSMENT');
    lines.push('═══════════════════════════════════════════════════════════════');
    const routingScore = results.routing.accuracy;
    const embeddingScore = (results.embedding.similarityAccuracy +
        results.embedding.searchMRR +
        results.embedding.clusterPurity) / 3;
    const overallScore = (routingScore + embeddingScore) / 2;
    lines.push('');
    lines.push(`  Routing Score:   ${(routingScore * 100).toFixed(1)}%`);
    lines.push(`  Embedding Score: ${(embeddingScore * 100).toFixed(1)}%`);
    lines.push(`  ─────────────────────────`);
    lines.push(`  Overall Score:   ${(overallScore * 100).toFixed(1)}%`);
    lines.push('');
    if (overallScore >= 0.8) {
        lines.push('  ✓ EXCELLENT - Highly suitable for Claude Code workflows');
    }
    else if (overallScore >= 0.6) {
        lines.push('  ~ GOOD - Suitable for most Claude Code use cases');
    }
    else if (overallScore >= 0.4) {
        lines.push('  ~ ACCEPTABLE - May work but consider alternatives');
    }
    else {
        lines.push('  ✗ NEEDS IMPROVEMENT - Consider different model or fine-tuning');
    }
    lines.push('');
    lines.push('═══════════════════════════════════════════════════════════════');
    return lines.join('\n');
}
/**
 * Compare two models
 */
function compareModels(results1, results2) {
    const lines = [];
    lines.push('');
    lines.push('╔═══════════════════════════════════════════════════════════════════════════╗');
    lines.push('║                       MODEL COMPARISON                                    ║');
    lines.push('╚═══════════════════════════════════════════════════════════════════════════╝');
    lines.push('');
    const metrics = [
        { name: 'Routing Accuracy', v1: results1.routing.accuracy, v2: results2.routing.accuracy },
        { name: 'Similarity Detection', v1: results1.embedding.similarityAccuracy, v2: results2.embedding.similarityAccuracy },
        { name: 'Search MRR', v1: results1.embedding.searchMRR, v2: results2.embedding.searchMRR },
        { name: 'Search NDCG', v1: results1.embedding.searchNDCG, v2: results2.embedding.searchNDCG },
        { name: 'Cluster Purity', v1: results1.embedding.clusterPurity, v2: results2.embedding.clusterPurity },
        { name: 'Routing Latency (ms)', v1: results1.routing.avgLatencyMs, v2: results2.routing.avgLatencyMs, lowerBetter: true },
    ];
    lines.push(`${'Metric'.padEnd(25)} ${results1.model.padEnd(15)} ${results2.model.padEnd(15)} Winner`);
    lines.push('─'.repeat(70));
    for (const m of metrics) {
        const val1 = m.lowerBetter ? m.v1 : m.v1;
        const val2 = m.lowerBetter ? m.v2 : m.v2;
        let winner;
        if (m.lowerBetter) {
            winner = val1 < val2 ? results1.model : val2 < val1 ? results2.model : 'tie';
        }
        else {
            winner = val1 > val2 ? results1.model : val2 > val1 ? results2.model : 'tie';
        }
        const v1Str = m.lowerBetter ? val1.toFixed(2) : (val1 * 100).toFixed(1) + '%';
        const v2Str = m.lowerBetter ? val2.toFixed(2) : (val2 * 100).toFixed(1) + '%';
        lines.push(`${m.name.padEnd(25)} ${v1Str.padEnd(15)} ${v2Str.padEnd(15)} ${winner}`);
    }
    return lines.join('\n');
}
