// @claude-flow/claims/dist/domain/events.js
/**
 * Claim Domain Events (ADR-007)
 *
 * Domain events for the claims system following event sourcing pattern.
 * All state changes emit events for audit trail and projections.
 *
 * @module v3/claims/domain/events
 */
// =============================================================================
// Event Factory Functions
// =============================================================================
let eventCounter = 0;
function generateEventId() {
    return `claim-evt-${Date.now()}-${++eventCounter}`;
}
function createClaimEvent(type, aggregateId, payload, metadata, causationId, correlationId) {
    return {
        id: generateEventId(),
        type,
        aggregateId,
        aggregateType: 'claim',
        version: 1, // Version will be set by event store
        timestamp: Date.now(),
        source: 'claim-service',
        payload,
        metadata,
        causationId,
        correlationId,
    };
}
// =============================================================================
// Public Event Factory Functions
// =============================================================================
export function createClaimCreatedEvent(claimId, issueId, claimant, expiresAt) {
    return createClaimEvent('claim:created', claimId, {
        claimId,
        issueId,
        claimant,
        claimedAt: Date.now(),
        expiresAt,
    });
}
export function createClaimReleasedEvent(claimId, issueId, claimant, reason) {
    return createClaimEvent('claim:released', claimId, {
        claimId,
        issueId,
        claimant,
        releasedAt: Date.now(),
        reason,
    });
}
export function createClaimExpiredEvent(claimId, issueId, claimant, lastActivityAt) {
    return createClaimEvent('claim:expired', claimId, {
        claimId,
        issueId,
        claimant,
        expiredAt: Date.now(),
        lastActivityAt,
    });
}
export function createClaimStatusChangedEvent(claimId, issueId, previousStatus, newStatus, note) {
    return createClaimEvent('claim:status-changed', claimId, {
        claimId,
        issueId,
        previousStatus,
        newStatus,
        changedAt: Date.now(),
        note,
    });
}
export function createClaimNoteAddedEvent(claimId, issueId, note, addedBy) {
    return createClaimEvent('claim:note-added', claimId, {
        claimId,
        issueId,
        note,
        addedAt: Date.now(),
        addedBy,
    });
}
export function createHandoffRequestedEvent(claimId, issueId, handoffId, from, to, reason) {
    return createClaimEvent('handoff:requested', claimId, {
        claimId,
        issueId,
        handoffId,
        from,
        to,
        reason,
        requestedAt: Date.now(),
    });
}
export function createHandoffAcceptedEvent(claimId, issueId, handoffId, from, to) {
    return createClaimEvent('handoff:accepted', claimId, {
        claimId,
        issueId,
        handoffId,
        from,
        to,
        acceptedAt: Date.now(),
    });
}
export function createHandoffRejectedEvent(claimId, issueId, handoffId, from, to, reason) {
    return createClaimEvent('handoff:rejected', claimId, {
        claimId,
        issueId,
        handoffId,
        from,
        to,
        rejectedAt: Date.now(),
        reason,
    });
}
export function createReviewRequestedEvent(claimId, issueId, reviewers, requestedBy) {
    return createClaimEvent('review:requested', claimId, {
        claimId,
        issueId,
        reviewers,
        requestedAt: Date.now(),
        requestedBy,
    });
}
export function createReviewCompletedEvent(claimId, issueId, reviewer, approved, comments) {
    return createClaimEvent('review:completed', claimId, {
        claimId,
        issueId,
        reviewer,
        approved,
        completedAt: Date.now(),
        comments,
    });
}
// =============================================================================
// Extended Event Factory Functions
// =============================================================================
export function createIssueMarkedStealableEvent(claimId, issueId, originalClaimant, reason, gracePeriodMs, minPriorityToSteal, requiresContest) {
    const now = Date.now();
    return {
        id: generateExtEventId(),
        type: 'steal:issue-marked-stealable',
        aggregateId: claimId,
        aggregateType: 'claim',
        version: 1,
        timestamp: now,
        source: 'work-stealing-service',
        payload: {
            claimId,
            issueId,
            originalClaimant,
            reason,
            gracePeriodMs,
            gracePeriodEndsAt: now + gracePeriodMs,
            minPriorityToSteal,
            requiresContest,
        },
    };
}
export function createIssueStolenExtEvent(claimId, issueId, originalClaimant, newClaimant, reason, hadContest, progressTransferred, contestId) {
    return {
        id: generateExtEventId(),
        type: 'steal:issue-stolen',
        aggregateId: claimId,
        aggregateType: 'claim',
        version: 1,
        timestamp: Date.now(),
        source: 'work-stealing-service',
        payload: {
            claimId,
            issueId,
            originalClaimant,
            newClaimant,
            reason,
            hadContest,
            contestId,
            progressTransferred,
        },
    };
}
export function createSwarmRebalancedExtEvent(claimsMoved, moves, loadBefore, loadAfter, durationMs, trigger, errors = []) {
    return {
        id: generateExtEventId(),
        type: 'swarm:rebalanced',
        aggregateId: 'swarm',
        aggregateType: 'claim',
        version: 1,
        timestamp: Date.now(),
        source: 'load-balancer',
        payload: {
            claimsMoved,
            moves,
            loadBefore,
            loadAfter,
            durationMs,
            trigger,
            errors,
        },
    };
}
export function createAgentOverloadedExtEvent(agentId, agentName, currentLoad, threshold, activeClaims, maxClaims, recommendedAction) {
    return {
        id: generateExtEventId(),
        type: 'agent:overloaded',
        aggregateId: agentId,
        aggregateType: 'claim',
        version: 1,
        timestamp: Date.now(),
        source: 'load-balancer',
        payload: {
            agentId,
            agentName,
            currentLoad,
            threshold,
            activeClaims,
            maxClaims,
            recommendedAction,
        },
    };
}
export function createAgentUnderloadedExtEvent(agentId, agentName, currentLoad, threshold, activeClaims, maxClaims, availableCapacity) {
    return {
        id: generateExtEventId(),
        type: 'agent:underloaded',
        aggregateId: agentId,
        aggregateType: 'claim',
        version: 1,
        timestamp: Date.now(),
        source: 'load-balancer',
        payload: {
            agentId,
            agentName,
            currentLoad,
            threshold,
            activeClaims,
            maxClaims,
            availableCapacity,
        },
    };
}
let extEventCounter = 0;
function generateExtEventId() {
    return `claim-ext-evt-${Date.now()}-${++extEventCounter}`;
}
